library/src/scala/collection/mutable/IArrayBuilder.scala

```diff
@@ -0,0 +1,192 @@
+package scala.collection.mutable
+
+import scala.reflect.ClassTag
+
+/** A builder class for immutable arrays.
+ *
+ *  @tparam T    the type of the elements for the builder.
+ */
+@SerialVersionUID(3L)
+sealed abstract class IArrayBuilder[T]
+  extends ReusableBuilder[T, IArray[T]]
+    with Serializable {
+
+  protected[this] val arrayBuilder: ArrayBuilder[T]
+
+  override final def sizeHint(size: Int): Unit = arrayBuilder.sizeHint(size)
+  def clear(): Unit = arrayBuilder.clear()
+
+  /** Add all elements of an array */
+  def addAll(xs: IArray[T]): this.type =
+    arrayBuilder.addAll(xs.asInstanceOf[Array[T]])
+    this
+
+  /** Add a slice of an array */
+  def addAll(xs: IArray[T], offset: Int, length: Int): this.type =
+    arrayBuilder.addAll(xs.asInstanceOf[Array[T]], offset, length)
+    this
+
+  override def addAll(xs: IterableOnce[T]): this.type =
+    arrayBuilder.addAll(xs)
+    this
+}
+
+/** A companion object for array builders. */
+object IArrayBuilder {
+
+  /** Creates a new arraybuilder of type `T`.
+  *
+  *  @tparam T     type of the elements for the array builder, with a `ClassTag` context bound.
+  *  @return       a new empty array builder.
+  */
+  @inline def make[T: ClassTag]: IArrayBuilder[T] = {
```

Comment:
I take it that using `mapResult` didn't work?

Reply:
It works, but the signature must be `Builder[T, IArray[T]]` and it adds an extra indirection. 

This would mean that we do not have the efficient `addAll(xs: IArray[T])` and `(xs: IArray[T], offset: Int, length: Int)` on that builder.

Reply:
I guess we could also have `addAll(xs: Array[T])`

Reply:
We have 2 options:
* Return `IArrayBuilder[T]`
  * Lareger code
  * Efficient `add` for `IArray`/`Array`
  * Possible to create a specialized `IArrayBuilder[T]` for primitive types
* Return `Builder[T, IArray[T]]`
  * Trivial code
  *  In-efficient `add` for `IArray`/`Array` 
  * Impossible to create a specialized `Builder[T, IArray[T]]` for primitive types

I'm inclined to have the `IArrayBuilder[T]` for performance reasons. Otherwise, we might see applications that use the `ArrayBuilder` and then cast it to `IArray` (or use `IArray.unsafeFromArray`)

Reply:
> Impossible to create a specialized Builder[T, IArray[T]] for primitive types

Hum, that's not the case. If you go through `ArrayBuilder.make[T]` followed by a `mapResult`, you will receive the specialized `ArrayBuilder`, which will create a specialized `Array[T]`, and the only thing is that is going to be cast to an `IArray[T]` at the end of the day, which even when not specialized is a no-op. So we would definitely have the specialized builders for `IArray`s.

Now we couldn't have `addAll` specialized for `IArray` and `Array`, that's true. However `addAll` is virtually never used on `Builder`s.

Reply:
The issue was alluding to is that the result of [`mapResult`](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/mutable/Builder.scala#L81) does not have an overload of `addAll` that receives Arrays.

But if it is almost never used I will remove this.